pkg/controller/beat/common/config.go

```diff
@@ -58,44 +67,137 @@ func setOutput(cfg *settings.CanonicalConfig, client k8s.Client, associated comm
 }
 
 func buildBeatConfig(
-	log logr.Logger,
-	client k8s.Client,
-	beat beatv1beta1.Beat,
+	params DriverParams,
 	defaultConfig *settings.CanonicalConfig,
 ) ([]byte, error) {
 	cfg := settings.NewCanonicalConfig()
 
-	if err := setOutput(cfg, client, &beat); err != nil {
+	if err := setOutput(cfg, params.Client, &params.Beat); err != nil {
 		return nil, err
 	}
 
-	// use only the default config or only the provided config - no overriding, no merging
-	userConfig := beat.Spec.Config
-	if userConfig == nil {
-		if err := cfg.MergeWith(defaultConfig); err != nil {
+	// get user config from `config` or `configRef`
+	userConfig, defaultUsed, err := getUserConfig(params, defaultConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// use only the default config or only the user provided config - no overriding, no merging
+	if userConfig != nil {
+		if err := cfg.MergeWith(userConfig); err != nil {
 			return nil, err
 		}
-	} else {
-		userCfg, err := settings.NewCanonicalConfigFrom(userConfig.Data)
-		if err != nil {
+		if !defaultUsed {
+			params.Logger.V(1).Info("Replacing default configuration by user-provided configuration")
+		}
+	} else if defaultConfig != nil {
+		if err := cfg.MergeWith(defaultConfig); err != nil {
 			return nil, err
 		}
+	}
+	return cfg.Render()
+}
 
-		if err = cfg.MergeWith(userCfg); err != nil {
-			return nil, err
+// getUserConfig extracts the config either from the spec `config` field or from the Secret referenced by spec
+// `configRef` field. If a Secret is referenced, but doesn't exist, it will be created and populated with the
+// default config. From that point on it will be treated the same as a user provided config - it will be watched
+// and changes will be propagated to the Beat config.
+// Returns the results config, bool to indicate whether config was created from default and error.
+func getUserConfig(params DriverParams, defaultConfig *settings.CanonicalConfig) (*settings.CanonicalConfig, bool, error) {
```

Comment:
I'm holding with that, because if we give up on presets/defaults (and it's likely that we do) we won't be creating Secrets on our own any more and this code will change.